### ISSUE TITLE
`reportUnknownVariableType` raised on `default_factory` with generic type

### ISSUE THREAD

**Issue by skilkis** [microsoft/pyright#11112]
**Describe the bug**
When in strict mode, if a dataclass field has a `default_factory` that is a generic type, a `reportUnknownVariableType` diagnostic message is raised even if the field is annotated with a concrete type. Utilizing a lambda seems to bypass this behavior or allow Pylance to infer the type from the field annotation.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
from dataclasses import dataclass, field
from typing import Any


# reportUnknownVariableType with Type of "<field>" is partially unknown:
@dataclass
class Foo:
    a: list[str] = field(default_factory=list)
    b: dict[str, Any] = field(default_factory=dict)


# Type-Checks Fine:
@dataclass
class Bar:
    a: list[str] = field(default_factory=lambda: [])
    b: dict[str, Any] = field(default_factory=lambda: {})
```

I would only expect this diagnostic for a field such as the following:

```python
a: list = field(default_factory=list)
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

> Python 3.13
> Pyright Playground (pyright 1.1.407) [link](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sAaKYfbVGAKFEigQE8j8A7AOYFipcgEE%2BXFtIDEUENhJkAqnwDWfMAHc%2BANTgh8cAEapsAFR7YoW-AgAWUS0WthgUAEQAeBkxgA%2BDwI8IgMEI1RULigAV3VNHQAuFgABeCQ0TAwWTKwoADEwMGSoUqg4RKhUfAwEAG1akABdKABeekZmAAoYbGA4GNQEAH1%2B5ARSLlbq2oBKFjKoY0qYfHGGhBA6CS4W9t9u3v7BkbGJkCnV8fnZJysAWgBhe2xkNTx8-mxktMpc7P%2BUAAQgYSmUKlUavVGnsOn4en0BkNRihzlN0BBjPBKnUmvNFstYGtoZttpJYQcYAjjsizpNpnBMdioABvAC%2B8yAA)
> VS Code Pylance: 2025.9.101

**Comment by erictraut** [comment]
Duplicate of https://github.com/microsoft/pyright/issues/10283.
